File: BrokerClient/ClientBrokerManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BrokerCommon;
using BrokerCommon.Models;

namespace BrokerClient
{

    public class ClientBrokerManager
    {
        internal SocketLayer client;
        private List<ClientPool> pools { get; } = new List<ClientPool>();
        private List<Swimmer> swimmers { get; } = new List<Swimmer>();
        public string MySwimmerId => client.Id;

        private Action onReady;
        private Action onDisconnect { get; set; }
        private OnMessage onMessage { get; set; }
        private OnMessageWithResponse onMessageWithResponse { get; set; }


        public void ConnectToBroker(string ip)
        {
            client = new SocketLayer("127.0.0.1", getSwimmerById);
            client.OnMessage += onReceiveMessage;
            client.OnMessageWithResponse += onReceiveMessageWithResponse;
            client.OnDisconnect += _ => onDisconnect?.Invoke();
            client.StartFromClient();
            GetSwimmerId((id) =>
            {
                client.Id = id;
                this.onReady?.Invoke();
            });
        }

        private Swimmer getSwimmerById(string id)
        {
            var swimmer = swimmers.FirstOrDefault(a => a.Id == id);
            if (swimmer == null)
            {
                swimmer = new Swimmer(client, id);
                swimmers.Add(swimmer);
            }
            return swimmer;
        }

        public void OnReady(Action callback)
        {
            this.onReady = callback;
        }
        public void OnDisconnect(Action callback)
        {
            this.onDisconnect = callback;
        }

        public void OnMessage(OnMessage callback)
        {
            onMessage += callback;
        }
        public void OnMessageWithResponse(OnMessageWithResponse callback)
        {
            onMessageWithResponse += callback;
        }

        private void onReceiveMessageWithResponse(Swimmer from, Query query, Action<Query> respond)
        {
            if (query.Contains("~ToSwimmer~"))
            {
                onMessageWithResponse?.Invoke(from, query, respond);
                return;
            }
            if (query.Contains("~ToPool~"))
            {
                var pool = pools.FirstOrDefault(a => a.PoolName == query["~ToPool~"]);
                pool?.onMessageWithResponse?.Invoke(from, query, respond);
                return;
            }
            if (query.Contains("~ToPoolAll~"))
            {
                var pool = pools.FirstOrDefault(a => a.PoolName == query["~ToPoolAll~"]);
                pool?.onMessageWithResponse?.Invoke(from, query, (res) =>
                {
                    res.Add("~PoolAllCount~", query["~PoolAllCount~"]);
                    respond(res);
                });
                return;
            }
        }



        private void onReceiveMessage(Swimmer from, Query query)
        {
            if (query.Contains("~ToSwimmer~"))
            {
                onMessage?.Invoke(from, query);
                return;
            }
            if (query.Contains("~ToPool~"))
            {
                var pool = pools.FirstOrDefault(a => a.PoolName == query["~ToPool~"]);
                pool?.onMessage.Invoke(from, query);
                return;
            }
            if (query.Contains("~ToPoolAll~"))
            {
                var pool = pools.FirstOrDefault(a => a.PoolName == query["~ToPoolAll~"]);
                pool?.onMessage.Invoke(from, query);
                return;
            }
        }


        public void GetSwimmerId(Action<string> callback)
        {
            var query = Query.Build("GetSwimmerId");

            client.SendMessageWithResponse(query, (response) =>
            {
                callback(response.GetJson<string>());
            });
        }


        public void SendMessage(string swimmerId, Query query)
        {
            query.Add("~ToSwimmer~", swimmerId);
            client.SendMessage(query);
        }

        public void SendMessageWithResponse(string swimmerId, Query query, Action<Query> callback)
        {
            query.Add("~ToSwimmer~", swimmerId);
            client.SendMessageWithResponse(query, callback);
        }

        public void GetPool(string poolName, Action<ClientPool> callback)
        {
            var pool = pools.FirstOrDefault(a => a.PoolName == poolName);
            if (pool != null)
            {
                callback(pool);
                return;
            }

            var query = Query.Build("GetPool", new QueryParam("PoolName", poolName));

            client.SendMessageWithResponse(query, (response) =>
            {
                var getPoolByNameResponse = response.GetJson<GetPoolByNameResponse>();
                pool = pools.FirstOrDefault(a => a.PoolName == getPoolByNameResponse.PoolName);
                if (pool == null)
                {
                    pools.Add(pool = new ClientPool(this, getPoolByNameResponse, getSwimmerById));
                }

                pool.PoolName = getPoolByNameResponse.PoolName;
                callback(pool);
            });
        }

        public void GetAllPools(string poolName, Action<GetAllPoolsResponse> callback)
        {
            var query = Query.Build("GetAllPools");

            client.SendMessageWithResponse(query, (response) =>
            {
                callback(response.GetJson<GetAllPoolsResponse>());
            });
        }

        public void Disconnet()
        {
            client.ForceDisconnect();
        }

    }
}

File: BrokeClientTS/src/clientBrokerManager.ts
import {ClientPool} from "./clientPool";
import {ClientConnection, OnMessage, OnMessageWithResponse} from "./common/clientConnection";
import {Query, QueryParam} from "./common/query";
import {GetPoolByNameResponse} from "./models/GetPoolByNameResponse";
import {GetAllPoolsResponse} from "./models/GetAllPoolsResponse";

export class ClientBrokerManager {
    private pools: ClientPool[] = [];
    public client: ClientConnection;

    public get MySwimmerId(): string {
        return this.client.Id;
    }

    private onReady: (() => void)[] = [];
    private onDisconnect: (() => void)[] = [];
    private onMessage: OnMessage[] = [];
    private onMessageWithResponse: OnMessageWithResponse[] = [];

    public ConnectToBroker(ip: string): void {
        this.client = new ClientConnection("127.0.0.1");
        this.client.OnMessage.push((from, message) => this.onReceiveMessage(from, message));
        this.client.OnMessageWithResponse.push((from, message, respond) => this.onReceiveMessageWithResponse(from, message, respond));
        this.client.OnDisconnect.push(_ => this.invokeDisconnect());
        this.client.StartFromClient();

        this.GetSwimmerId((id) => {
            this.client.Id = id;
            this.invokeReady();
        });
    }

    public OnReady(callback: () => void): void {
        this.onReady.push(callback);
    }

    public OnDisconnect(callback: () => void): void {
        this.onDisconnect.push(callback);
    }

    public OnMessage(call: OnMessage): void {
        this.onMessage.push(call);
    }

    public OnMessageWithResponse(callback: OnMessageWithResponse): void {
        this.onMessageWithResponse.push(callback);
    }

    private onReceiveMessageWithResponse(from: ClientConnection, message: Query, respond: (query: Query) => void): void {
        if (message.Contains("~ToSwimmer~")) {
            this.invokeMessageWithResponse(from, message, respond);
            return
        }
        if (message.Contains("~ToPool~")) {
            let pool = this.pools.filter(a => a.PoolName == message.get("~ToPool~"))[0];
            pool && pool.invokeMessageWithResponse(from, message, respond);
            return;
        }
        if (message.Contains("~ToPoolAll~")) {
            let pool = this.pools.filter(a => a.PoolName == message.get("~ToPoolAll~"))[0];
            pool &&
            pool.invokeMessageWithResponse(from, message,
                (res) => {
                    res.Add("~PoolAllCount~", message.get("~PoolAllCount~"));
                    respond(res);
                });
            return;
        }
    }

    private onReceiveMessage(from: ClientConnection, message: Query): void {
        if (message.Contains("~ToSwimmer~")) {
            this.invokeMessage(from, message);
            return
        }
        if (message.Contains("~ToPool~")) {
            let pool = this.pools.filter(a => a.PoolName == message.get("~ToPool~"))[0];
            pool && pool.invokeMessage(from, message);
            return;
        }
        if (message.Contains("~ToPoolAll~")) {
            let pool = this.pools.filter(a => a.PoolName == message.get("~ToPoolAll~"))[0];
            pool && pool.invokeMessage(from, message);
            return;
        }
    }

    public GetSwimmerId(callback: (_: string) => void): void {
        var query = Query.Build("GetSwimmerId");
        this.client.SendMessageWithResponse(query,
            (response) => {
                callback(response.GetJson<string>());
            });
    }
    public SendMessage(swimmerId: string, query: Query): void {
        query.Add("~ToSwimmer~", swimmerId);
        this.client.SendMessage(query);
    }
    public SendMessageWithResponse<T>(swimmerId: string, query: Query, callback: (response: T) => void): void {
        query.Add("~ToSwimmer~", swimmerId);
        this.client.SendMessageWithResponse(query, (response) => {
            callback(response.GetJson<T>());
        });
    }
    public GetPool(poolName: string, callback: (_: ClientPool) => void): void {
        var query = Query.Build("GetPool", new QueryParam("PoolName", poolName));
        this.client.SendMessageWithResponse(query,
            (response) => {
                var getPoolByNameResponse = response.GetJson<GetPoolByNameResponse>();
                var pool: ClientPool = this.pools.filter(a => a.PoolName == getPoolByNameResponse.PoolName)[0];
                if (pool == null) {
                    this.pools.push(pool = new ClientPool(this, getPoolByNameResponse));
                }
                pool.PoolName = getPoolByNameResponse.PoolName;
                callback(pool);
            });
    }

    public GetAllPools(poolName: string, callback: (_: GetAllPoolsResponse) => void): void {
        var query = Query.Build("GetAllPools");
        this.client.SendMessageWithResponse(query,
            (response) => {
                callback(response.GetJson<GetAllPoolsResponse>());
            });
    }

    public Disconnet(): void {
        this.client.ForceDisconnect();
    }

    private invokeMessageWithResponse(from: ClientConnection, message: Query, respond: (query: Query) => void) {
        for (var i = 0; i < this.onMessageWithResponse.length; i++) {
            this.onMessageWithResponse[i](from, message, respond);
        }
    }

    private invokeMessage(from: ClientConnection, message: Query) {
        for (var i = 0; i < this.onMessage.length; i++) {
            this.onMessage[i](from, message);
        }
    }

    private invokeDisconnect() {
        for (var i = 0; i < this.onDisconnect.length; i++) {
            this.onDisconnect[i]();
        }
    }

    private invokeReady() {
        for (var i = 0; i < this.onReady.length; i++) {
            this.onReady[i]();
        }
    }

}